### DISCUSSION TITLE
docs/web/router-middleware-intro

### DISCUSSION THREAD

**Discussion by giscus** [gogf/gf-site#80] [General]
# docs/web/router-middleware-intro

GoFrame框架中中间件和拦截器的基本使用。GoFrame通过优雅的中间件请求控制方式和经典的中间件洋葱模型，为WebServer提供灵活强大的插件机制。讨论了中间件定义、类型、注册方式及执行优先级，以提供全面的请求流程控制方案。

https://goframe.org/docs/web/router-middleware-intro

**Comment by Starix610** [comment]
想要实现一个中间件，针对指定路由来做err检测告警，如何拿到每个路由处理后的err信息呢

**Reply by gqcn** [reply]
可以使用`r.GetError()`获取到处理流程产生的`error`，包括业务函数执行`return`的`error`，具体可以看一下章节：https://goframe.org.cn/docs/web/panic-handling#%E7%9B%B8%E5%85%B3%E6%96%B9%E6%B3%95

**Comment by i-bing** [comment]
自定义中间件，如何单测中构造*ghttp.Request ？
由于中间件依赖 *ghttp.Request ， 发现newRequest 没有暴露，手动new ，发现middleware struct 也没有暴露，r.Middleware.Next() 会panic
